### PR TITLE
Move writes to Transaction

### PR DESCRIPTION
**Pull request by syun64**
As a followup from @HonahX 's suggestion on https://github.com/apache/iceberg-python/pull/498

This refactoring is required both for supporting CreateTableTransaction, and supporting static overwrites through delete + append

### CODE REVIEW

File: tests/integration/test_writes.py [hunk 1]
```diff
@@ -778,3 +778,31 @@ def test_inspect_snapshots(
                 continue
 
             assert left == right, f"Difference in column {column}: {left} != {right}"
+
+
+@pytest.mark.integration
+def test_write_within_transaction(spark: SparkSession, session_catalog: Catalog, arrow_table_with_null: pa.Table) -> None:
```
Comment (jqin61) [inline tests/integration/test_writes.py:784]: Shall we cover overwrite in the test as well?

Comment (syun64) [reply tests/integration/test_writes.py:784]: Sure thing :)

**Review by HonahX — APPROVED** [review]
LGTM! Thanks @syun64 for taking this and @jqin61 for reviewing!

Continue the discussion https://github.com/apache/iceberg-python/pull/498#issuecomment-2033344103 here:

I agree that removing them from Table class can simplify the implementation and emphasize the concept of "transaction". It appears that these methods in the Table class serve as handy tools for users to perform simple updates or data insertion. Considering these methods were released in 0.6.0, how about we initiate an issue to gather feedback from our users? This could facilitate a discussion on whether these methods should be deprecated.

**Comment by Fokko** [comment]
The methods on the table were added as shorthands indeed. I'm open to removing those methods and letting everything go through a transaction. This way we have a single way to do things, and it also emphasizes that it is a transaction and done in an atomic way.

**Review by Fokko — APPROVED** [review]

**Comment by Fokko** [comment]
Thanks for working on this @syun64 and @HonahX for the review 🙌

**Pull request merged**